kanidmd/src/lib/idm/authsession.rs

```diff
@@ -309,8 +310,22 @@ impl CredHandler {
                             CredState::Denied(BAD_TOTP_MSG)
                         }
                     }
-                    (AuthCredential::BackupCode(code_chal), _, _, Some(backup_code)) => {
-                        if backup_code.verify(&code_chal) {
+                    (AuthCredential::BackupCode(code_chal), _, _, Some(backup_codes)) => {
+                        // Clone backup_codes because it's not mutable. If I make it a mutable ref,
+                        // then I have to pass it as &BackupCodes and specify lifetime on BackupCodeRemoval
+                        let mut cloned_codes = backup_codes.clone();
+                        if cloned_codes.verify(&code_chal) {
+                            if let Err(_e) =
+                                async_tx.send(DelayedAction::BackupCodeRemoval(BackupCodeRemoval {
+                                    target_uuid: who,
+                                    updated_codes: cloned_codes,
```

Comment:
There is a problem here. Imagine we have two logins, and they each use a seperate backup code, A and B. We'll have two delayed actions queued. One set where A is NOT present, one where B is not present. So we'll "lose" one of the code removals. 

Also imagine the person reset's their backup codes, before this delayed action is processed. We would then process his and "reset back" all their codes! 

The correct action here is not to clone the codes to send in the delayed action, but to send the *code that was used*. and then the delayed action looks at the set within a write txn and removes that code, if present, or if not present, skips (because the codes were reset). 

This avoids your cloning issue too :) 

tl;dr - 
```
BackupCodeRemoval {
    target_uuid: Uuid,
    code_to_remove: String
}
```

Reply:
Good catch. So when the delayed action is processed, I should retrieve the backup codes and remove `code_to_remove` from it, right?

Reply:
Correct! :D